File: parser/src/graph.ts
import { AlpObject } from './reader';
import { ValidationError } from './error';

export interface GraphNode {
  id: string;
  type: string;
  object: AlpObject;
}

export interface GraphEdge {
  source: string;
  target: string;
  type: 'blocks' | 'requires' | 'extends' | 'uses' | 'implements' | 'references';
}

/**
 * The ALP Graph Engine.
 * Builds a Directed Acyclic Graph (DAG) from parsed ALP objects,
 * resolves references, detects cycles, and enables topological sorting.
 */
export class AlpGraph {
  public nodes: Map<string, GraphNode> = new Map();
  public edges: GraphEdge[] = [];

  /**
   * Build the graph from a list of parsed ALP objects.
   * Extracts nodes (objects with IDs) and edges (-> ref properties).
   */
  public buildGraph(objects: AlpObject[]): void {
    this.nodes.clear();
    this.edges = [];

    // Phase 1: Register all nodes
    for (const obj of objects) {
      if (obj.id) {
        this.nodes.set(obj.id, {
          id: obj.id,
          type: obj._type,
          object: obj,
        });
      }
    }

    // Phase 2: Extract edges from reference properties
    for (const obj of objects) {
      if (!obj.id) continue;

      for (const [key, value] of Object.entries(obj)) {
        if (key === '_type' || key === 'id') continue;

        if (typeof value === 'string' && value.startsWith('-> ')) {
          const targetId = value.substring(3).trim();
          const edgeType = this.inferEdgeType(key);
          this.edges.push({
            source: obj.id,
            target: targetId,
            type: edgeType,
          });
        }

        // Handle arrays of references (e.g., depends_on list)
        if (Array.isArray(value)) {
          const edgeType = this.inferEdgeType(key);
          for (const item of value) {
            if (typeof item === 'string' && item.startsWith('-> ')) {
              const targetId = item.substring(3).trim();
              this.edges.push({
                source: obj.id,
                target: targetId,
                type: edgeType,
              });
            }
          }
        }
      }
    }
  }

  /**
   * Detect cycles in the graph using DFS.
   * Throws a ValidationError if a cycle is found.
   */
  public detectCycles(): void {
    const visited = new Set<string>();
    const recursionStack = new Set<string>();
    const path: string[] = [];

    const dfs = (nodeId: string): void => {
      visited.add(nodeId);
      recursionStack.add(nodeId);
      path.push(nodeId);

      const outgoing = this.edges.filter(
        e => e.source === nodeId && (e.type === 'blocks' || e.type === 'requires')
      );
      for (const edge of outgoing) {
        if (!visited.has(edge.target)) {
          dfs(edge.target);
        } else if (recursionStack.has(edge.target)) {
          const cycleStart = path.indexOf(edge.target);
          const cycle = [...path.slice(cycleStart), edge.target];
          throw new ValidationError(
            `Dependency cycle detected: ${cycle.join(' → ')}`
          );
        }
      }

      path.pop();
      recursionStack.delete(nodeId);
    };

    for (const nodeId of this.nodes.keys()) {
      if (!visited.has(nodeId)) {
        dfs(nodeId);
      }
    }
  }

  /**
   * Return nodes in topological order (execution order).
   * Only considers 'blocks' and 'requires' edges.
   */
  public topologicalSort(): GraphNode[] {
    const orderingEdges = this.edges.filter(
      e => e.type === 'blocks' || e.type === 'requires'
    );

    // Build out-degree map (how many things a node depends on)
    const outDegree = new Map<string, number>();
    for (const nodeId of this.nodes.keys()) {
      outDegree.set(nodeId, 0);
    }
    for (const edge of orderingEdges) {
      if (outDegree.has(edge.source)) {
        outDegree.set(edge.source, (outDegree.get(edge.source) || 0) + 1);
      }
    }

    // Kahn's algorithm (reversed for dependency resolution)
    const queue: string[] = [];
    for (const [nodeId, degree] of outDegree.entries()) {
      if (degree === 0) queue.push(nodeId);
    }

    const sorted: GraphNode[] = [];
    while (queue.length > 0) {
      const current = queue.shift()!;
      const node = this.nodes.get(current);
      if (node) sorted.push(node);

      // We completed 'current'. Anyone who depends on 'current' has one less dependency.
      for (const edge of orderingEdges) {
        if (edge.target === current && outDegree.has(edge.source)) {
          const newDegree = (outDegree.get(edge.source) || 1) - 1;
          outDegree.set(edge.source, newDegree);
          if (newDegree === 0) queue.push(edge.source);
        }
      }
    }

    return sorted;
  }

  /**
   * Get all downstream nodes affected by a change to the given node.
   */
  public getImpact(nodeId: string): GraphNode[] {
    const impacted = new Set<string>();
    const queue = [nodeId];

    while (queue.length > 0) {
      const current = queue.shift()!;
      const dependents = this.edges
        .filter(e => e.target === current)
        .map(e => e.source);

      for (const dep of dependents) {
        if (!impacted.has(dep) && dep !== nodeId) {
          impacted.add(dep);
          queue.push(dep);
        }
      }
    }

    return Array.from(impacted)
      .map(id => this.nodes.get(id))
      .filter((n): n is GraphNode => n !== undefined);
  }

  /**
   * Get all upstream nodes that must complete before the given node can start.
   */
  public getBlockers(nodeId: string): GraphNode[] {
    const blockers = new Set<string>();
    const queue = [nodeId];

    while (queue.length > 0) {
      const current = queue.shift()!;
      const dependencies = this.edges
        .filter(e => e.source === current && (e.type === 'blocks' || e.type === 'requires'))
        .map(e => e.target);

      for (const dep of dependencies) {
        if (!blockers.has(dep) && dep !== nodeId) {
          blockers.add(dep);
          queue.push(dep);
        }
      }
    }

    return Array.from(blockers)
      .map(id => this.nodes.get(id))
      .filter((n): n is GraphNode => n !== undefined);
  }

  /**
   * Print the graph as a readable tree structure.
   */
  public toTextTree(): string {
    const lines: string[] = [];
    const sorted = this.topologicalSort();

    for (const node of sorted) {
      const deps = this.edges
        .filter(e => e.source === node.id)
        .map(e => `→ ${e.target} (${e.type})`);

      const status = node.object.status || '';
      lines.push(`${status} @${node.type} ${node.id}`);
      for (const dep of deps) {
        lines.push(`    ${dep}`);
      }
    }

    return lines.join('\n');
  }

  /**
   * Infer the edge type from the property name.
   */
  private inferEdgeType(key: string): GraphEdge['type'] {
    switch (key) {
      case 'depends_on':
      case 'blocked_by':
        return 'blocks';
      case 'requires':
        return 'requires';
      case 'extends':
        return 'extends';
      case 'uses':
        return 'uses';
      case 'implements':
        return 'implements';
      default:
        return 'references';
    }
  }
}

File: sdk/python/alp_sdk/graph.py
"""ALP graph engine (v6.4.0 - Python SDK parity).

Mirrors the TypeScript ``@alp/parser`` ``AlpGraph``: builds a DAG from parsed
ALP objects, resolves ``-> ref`` edges, detects cycles, and supports
topological sort plus impact/blocker queries.
"""

from typing import Any, Dict, List, Optional, Set, Tuple

from .models import AlpObject


class GraphNode:
    def __init__(self, id: str, type: str, object: AlpObject):
        self.id = id
        self.type = type
        self.object = object

    def __repr__(self) -> str:
        return f"GraphNode(id={self.id!r}, type={self.type!r})"


class GraphEdge:
    TYPES = ("blocks", "requires", "extends", "uses", "implements", "references")

    def __init__(self, source: str, target: str, type: str):
        self.source = source
        self.target = target
        self.type = type

    def __repr__(self) -> str:
        return f"GraphEdge({self.source} -{self.type}-> {self.target})"


def _infer_edge_type(key: str) -> str:
    if key in ("depends_on", "blocked_by"):
        return "blocks"
    if key == "requires":
        return "requires"
    if key == "extends":
        return "extends"
    if key == "uses":
        return "uses"
    if key == "implements":
        return "implements"
    return "references"


class AlpGraph:
    """Build and query the ALP dependency graph."""

    def __init__(self) -> None:
        self.nodes: Dict[str, GraphNode] = {}
        self.edges: List[GraphEdge] = []

    def build_graph(self, objects: List[AlpObject]) -> None:
        self.nodes.clear()
        self.edges = []

        for obj in objects:
            if obj.id:
                self.nodes[obj.id] = GraphNode(obj.id, obj._type, obj)

        for obj in objects:
            if not obj.id:
                continue
            for key, value in obj.properties.items():
                if key in ("_type", "id"):
                    continue
                if isinstance(value, str) and value.startswith("-> "):
                    target = value[3:].strip()
                    self.edges.append(
                        GraphEdge(obj.id, target, _infer_edge_type(key))
                    )
                elif isinstance(value, list):
                    etype = _infer_edge_type(key)
                    for item in value:
                        if isinstance(item, str) and item.startswith("-> "):
                            target = item[3:].strip()
                            self.edges.append(GraphEdge(obj.id, target, etype))

    def detect_cycles(self) -> None:
        """Raise ``ValueError`` if a cycle exists among blocks/requires edges."""
        visited: Set[str] = set()
        stack: Set[str] = set()
        path: List[str] = []

        def dfs(node_id: str) -> None:
            visited.add(node_id)
            stack.add(node_id)
            path.append(node_id)
            for edge in self.edges:
                if edge.source == node_id and edge.type in ("blocks", "requires"):
                    if edge.target not in visited:
                        dfs(edge.target)
                    elif edge.target in stack:
                        start = path.index(edge.target)
                        cycle = path[start:] + [edge.target]
                        raise ValueError(
                            "Dependency cycle detected: " + " → ".join(cycle)
                        )
            path.pop()
            stack.discard(node_id)

        for node_id in list(self.nodes.keys()):
            if node_id not in visited:
                dfs(node_id)

    def topological_sort(self) -> List[GraphNode]:
        """Return nodes in execution order (Kahn's algorithm)."""
        ordering = [e for e in self.edges if e.type in ("blocks", "requires")]
        out_degree: Dict[str, int] = {nid: 0 for nid in self.nodes}
        for edge in ordering:
            if edge.source in out_degree:
                out_degree[edge.source] += 1

        queue: List[str] = [nid for nid, deg in out_degree.items() if deg == 0]
        sorted_nodes: List[GraphNode] = []
        while queue:
            current = queue.pop(0)
            node = self.nodes.get(current)
            if node:
                sorted_nodes.append(node)
            for edge in ordering:
                if edge.target == current and edge.source in out_degree:
                    out_degree[edge.source] -= 1
                    if out_degree[edge.source] == 0:
                        queue.append(edge.source)
        return sorted_nodes

    def get_impact(self, node_id: str) -> List[GraphNode]:
        """All downstream nodes affected by a change to ``node_id``."""
        impacted: Set[str] = set()
        queue = [node_id]
        while queue:
            current = queue.pop(0)
            for edge in self.edges:
                if edge.target == current and edge.source != node_id:
                    if edge.source not in impacted:
                        impacted.add(edge.source)
                        queue.append(edge.source)
        return [n for n in (self.nodes.get(i) for i in impacted) if n is not None]

    def get_blockers(self, node_id: str) -> List[GraphNode]:
        """All upstream nodes that must complete before ``node_id``."""
        blockers: Set[str] = set()
        queue = [node_id]
        while queue:
            current = queue.pop(0)
            for edge in self.edges:
                if edge.source == current and edge.type in ("blocks", "requires"):
                    if edge.target != node_id and edge.target not in blockers:
                        blockers.add(edge.target)
                        queue.append(edge.target)
        return [n for n in (self.nodes.get(i) for i in blockers) if n is not None]

    def to_text_tree(self) -> str:
        lines: List[str] = []
        for node in self.topological_sort():
            status = node.object.properties.get("status", "")
            lines.append(f"{status} @{node.type} {node.id}")
            for edge in self.edges:
                if edge.source == node.id:
                    lines.append(f"    → {edge.target} ({edge.type})")
        return "\n".join(lines)
